components/script/dom/globalscope.rs

```diff
@@ -2939,6 +2948,31 @@ impl GlobalScope {
         self.gpu_id_hub.clone()
     }
 
+    pub fn add_gpu_device(&self, device: &GPUDevice) {
+        self.gpu_devices
```

Comment:
when are these cleaned up?

Reply:
These aren't. I still need to figure this out. Maybe somehow send the pipeline id with FreeDeviceId msg. Otherwise searvh for the DeviceId in all Pipeline(but that would be highly inefficient and difficult once we start support for Workers)

Reply:
If we want a device to be cleaned at the time `FreeDeviceId` is sent to the content process, we can send a `FreeDeviceId` to the server itself, so that `WGPU` removes the device from its hashmap and send the content process another message with the pipeline_id which could then remove the `GPUDevice` from above HashMap.

We'll have to follow this whole process only because we can't send the pipeline_id with `FreeDeviceId` from `IdentityRecycler` in `webgpu/identity.rs`. If we can figure that out somehow things would be much easier.
(But we'll still have to send a message to the server itself to remove `Device` from HashMap in `WGPU`)